static/js/user_groups.ts

```diff
@@ -153,3 +153,26 @@ export function get_user_groups_of_user(user_id: number): UserGroup[] {
     );
     return groups_of_user;
 }
+
+export function get_recursive_subgroups(user_group_id: number): Set<number> | undefined {
+    const user_group = user_group_by_id_dict.get(user_group_id);
+    if (user_group === undefined) {
+        blueslip.error(`Could not find user group with ID ${user_group_id}`);
+        return undefined;
+    }
+
+    const subgroup_ids = new Set(user_group.subgroups);
+
+    for (const subgroup_id of subgroup_ids) {
+        const user_group = user_group_by_id_dict.get(subgroup_id);
+        if (user_group === undefined) {
+            blueslip.error(`Could not find user group with ID ${subgroup_id}`);
+            return undefined;
+        }
+
+        for (const group_id of user_group.subgroups) {
+            subgroup_ids.add(group_id);
```

Comment:
So, the semantics of adding something to a `set` while iterating through it are not always well-defined in programming languages; do you know how JavaScript handles that case? I would prefer a design that is correct regardless of how browsers implement that detail, but barring that, a comment making clear it's correct.

Reply:
I tried some print debugging and it seems that the added element is visited in the loop as expected. The [documentation here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach) mentions that using `forEach` will gurantee that the added element will be visited. I cannot find any robust source for `for..of` syntax that I have used here but a pretty old [Stackoverflow thread](https://stackoverflow.com/questions/28306756/is-it-safe-to-delete-elements-in-a-set-while-iterating-with-for-of#:~:text=Yes%2C%20it%20is%20perfectly%20fine,a%20set%20while%20iterating%20it.) mentions that we can do this.

Reply:
The point here is about readability, not whether it works -- if it requires that kind of googling to determine that it's correct, then folks reading this will worry whether it is actually correct. I'll think about whether we want a comment or to actually change it.

Reply:
I added a comment.